### ISSUE TITLE
Problem with installing WLEDMM on Adafruit Matrix Portal S3

### ISSUE THREAD

**Issue by letmoc** [MoonModules/WLED#205]
### What happened?

Hello, I received two Adafruit Matrix Portal S3. I then flashed the current WLED-MM binary. Adafruit MatrixPortal -S3, with HUB75 support: WLEDMM_0.14.1-b32.40_adafruit_matrixportal_esp32s3.bin.

Unfortunately, the 64x64 matrix is very delayed, and the colors don’t really match either. 

### To Reproduce Bug

Always occurs. Adafruit Matrix Portal S3

### Expected Behavior

Does anyone have an idea what could be the issue?

### Install Method

From MoonModules Release Page

### What version/release of MM WLED?

WLEDMM_0.14.1-b32.40_adafruit_matrixportal_esp32s3.bin

### Which microcontroller/board are you seeing the problem on?

Other

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by letmoc** [comment]
Video:
https://wled.discourse.group/t/problem-with-hub75-adafruit-matrix-portal-s3/13630

**Comment by softhack007** [comment]
hi @letmoc, did you try the latest build? 
Its not released yet, but already very stable so you should give it a try https://github.com/MoonModules/WLED/actions/runs/12320735208

The color differences might be due to changes on the panel manufacturer side - its quite normal that panels purchased at different times (even from same supplier) will not match colors.

**Comment by letmoc** [comment]
Hi @softhack007 
Thank you for your feedback. I will test the firmware later. The displays are both the same; when you swap them, the color differences also shift. I will provide feedback later.

**Comment by letmoc** [comment]
@softhack007

Update failed!
Please check your file and retry!

An update fails. When I flash the .bin firmware using esp-flasher, it goes through, but afterwards the WLED-AP is not displayed. The board is not working. Firmware: WLEDMM_14.5.0-beta_adafruit_matrixportal_esp32s3.bin

**Comment by softhack007** [comment]
> The board is not working. Firmware: WLEDMM_14.5.0-beta_adafruit_matrixportal_esp32s3.bin

@netmindz can you test the file? I don't have a matrixportal board.

**Comment by softhack007** [comment]
> using esp-flasher

@letmoc is this "esp home flasher"? I yes, then please use a different tool. ESP home flasher creates broken installs.
[WLED ESP Flasher](https://kno.wled.ge/basics/install-wled-flasher) should work better.

**Comment by softhack007** [comment]
Another option for flashing:
Step 1. Use this web-based installer to flash any S3 (8MB with qspi psram) firmware: https://wled-install.github.io/
Step 2. OTA update to the adafruit_matrixportal firmware

**Comment by letmoc** [comment]
@softhack007

MoonMod WLED Software Update
WLEDMM_0.14.1-b32.40 matrixportal_esp32s3.bin
(ESP32-S3 build 2406240)

MoonMod WLED Software Update
Updating...
Please do not close or refresh the page :)

Update failed!
Please check your file and retry!

++++++++++++++++++++++++++++++++++++++

The OTA update fails. I used the WLED ESP Flasher, and after flashing, the board does not start, and no WLED-AP is displayed.

**Comment by letmoc** [comment]
Using 'COM7' as serial port.
Connecting...
Detecting chip type... ESP32-S3
Connecting...

Chip Info:
 - Chip Family: ESP32
 - Chip Model: ESP32-S3
 - Number of Cores: 1
 - Max CPU Frequency: 80MHz
 - Has Bluetooth: NO
 - Has Embedded Flash: NO
 - Has Factory-Calibrated ADC: NO
 - MAC Address: XX:XX:XX:XX:XX:XX
Uploading stub...
Running stub...
Stub running...
Changing baud rate to 460800
Changed.
 - Flash Size: 8MB
 - Flash Mode: dio
 - Flash Frequency: 80MHz
Erasing flash (this may take a while)...
Chip erase completed successfully in 5.9s
Flash will be erased from 0x00000000 to 0x00003fff...
Flash will be erased from 0x00008000 to 0x00008fff...
Flash will be erased from 0x0000e000 to 0x0000ffff...
Flash will be erased from 0x00010000 to 0x001a9fff...
Flash params set to 0x023f
Compressed 13936 bytes to 9695...
Writing at 0x00000000... (100 %)
Wrote 13936 bytes (9695 compressed) at 0x00000000 in 0.2 seconds (effective 732.1 kbit/s)...
Hash of data verified.
Compressed 3072 bytes to 128...
Writing at 0x00008000... (100 %)
Wrote 3072 bytes (128 compressed) at 0x00008000 in 0.0 seconds (effective 894.9 kbit/s)...
Hash of data verified.
Compressed 8192 bytes to 47...
Writing at 0x0000e000... (100 %)
Wrote 8192 bytes (47 compressed) at 0x0000e000 in 0.1 seconds (effective 1284.2 kbit/s)...
Hash of data verified.
Compressed 1675872 bytes to 1136744...
Writing at 0x00010000... (1 %)
Writing at 0x00019165... (2 %)
Writing at 0x0001e1b5... (4 %)
Writing at 0x00022235... (5 %)
Writing at 0x0002624d... (7 %)
Writing at 0x0002a263... (8 %)
Writing at 0x0002e25c... (10 %)
Writing at 0x000321d8... (11 %)
Writing at 0x0003b35e... (12 %)
Writing at 0x00046b80... (14 %)
Writing at 0x0004f045... (15 %)
Writing at 0x00060659... (17 %)
Writing at 0x00065a12... (18 %)
Writing at 0x0006aab9... (20 %)
Writing at 0x0006fd5c... (21 %)
Writing at 0x00074cd8... (22 %)
Writing at 0x0007a05f... (24 %)
Writing at 0x0007fd1f... (25 %)
Writing at 0x000853cc... (27 %)
Writing at 0x0008a655... (28 %)
Writing at 0x0009076d... (30 %)
Writing at 0x00096298... (31 %)
Writing at 0x0009c224... (32 %)
Writing at 0x000a2782... (34 %)
Writing at 0x000a8059... (35 %)
Writing at 0x000ad330... (37 %)
Writing at 0x000b2dd9... (38 %)
Writing at 0x000b8a4a... (40 %)
Writing at 0x000bdd4b... (41 %)
Writing at 0x000c333f... (42 %)
Writing at 0x000c9639... (44 %)
Writing at 0x000cf638... (45 %)
Writing at 0x000d51b7... (47 %)
Writing at 0x000db876... (48 %)
Writing at 0x000e1811... (50 %)
Writing at 0x000e71e9... (51 %)
Writing at 0x000ed776... (52 %)
Writing at 0x000f36fb... (54 %)
Writing at 0x000f907e... (55 %)
Writing at 0x000fec0b... (57 %)
Writing at 0x00104714... (58 %)
Writing at 0x00109d71... (60 %)
Writing at 0x0010fbc1... (61 %)
Writing at 0x00114d9d... (62 %)
Writing at 0x00119b91... (64 %)
Writing at 0x0011eb03... (65 %)
Writing at 0x00123ce6... (67 %)
Writing at 0x00128be3... (68 %)
Writing at 0x0012db6e... (70 %)
Writing at 0x00132d8d... (71 %)
Writing at 0x00137a36... (72 %)
Writing at 0x0013cb16... (74 %)
Writing at 0x0014224f... (75 %)
Writing at 0x001479f8... (77 %)
Writing at 0x0014c726... (78 %)
Writing at 0x0015185e... (80 %)
Writing at 0x001567aa... (81 %)
Writing at 0x0015bc7e... (82 %)
Writing at 0x00161068... (84 %)
Writing at 0x00166a5b... (85 %)
Writing at 0x0016c34b... (87 %)
Writing at 0x001718cb... (88 %)
Writing at 0x00176d9e... (90 %)
Writing at 0x0017f186... (91 %)
Writing at 0x001877c5... (92 %)
Writing at 0x0018d7ab... (94 %)
Writing at 0x001963c0... (95 %)
Writing at 0x0019b816... (97 %)
Writing at 0x001a1148... (98 %)
Writing at 0x001a6db0... (100 %)
Wrote 1675872 bytes (1136744 compressed) at 0x00010000 in 10.7 seconds (effective 1252.6 kbit/s)...
Hash of data verified.

Leaving...
Hard Resetting...
Hard resetting via RTS pin...
Done! Flashing is complete!

**Comment by letmoc** [comment]
![IMG_2242](https://github.com/user-attachments/assets/c29224c6-16ca-47e9-9ee0-d79a051f26f0)

**Comment by letmoc** [comment]
@softhack007
For your information, I successfully updated the WF2 board to the BETA version via OTA. That works.

**Comment by softhack007** [comment]
Ok, I did some research. It seems that adafruit wants it "their way", so its best to use their tools and use their stock bootloader.

1. follow adafruit instructions to install and run the factory-reset-and-bootloader firmware: https://learn.adafruit.com/adafruit-matrixportal-s3/factory-reset#factory-reset-and-bootloader-repair-3107941
    (use the adafruit web flasher aka "WebSerial ESPTool Method")

2. unzip the attached files, and flash them with the adafruit web flasher https://adafruit.github.io/Adafruit_WebSerial_ESPTool
   * offset `0x8000` partitions_8MB_tinyuf2.bin
   * offset `0x10000` WLEDMM_14.5.0-beta_matrixportal_esp32s3_tinyuf2.bin
   * hit "Program"

3. restart your matrixportal board. This should give you a working WLED-MM installation. At first boot, it may take 1-3 minutes before WLED-AP appears, because WLED needs some time to prepare the flash filesystem.

"should" because I don't have the matrixportal board to test.

<br/>

--> After this procedure (all 3 steps, no cutting corners), you _should_ be able to OTA update to any (newer) version of WLED or WLED-MM.

[matrixportal_tinyUF2.zip](https://github.com/user-attachments/files/18197561/matrixportal_tinyUF2.zip)

**Comment by letmoc** [comment]
You are the best! 
It works now. 
The matrix behaves identically now, and the colors are as they should be.

**Comment by softhack007** [comment]
@letmoc thanks for reporting the problem - it looks like we should update our HUB75 documentation for matrixportal.

**Comment by letmoc** [comment]
@softhack007 
Thank you for your quick help. 
I also flashed the second board in this way. It worked immediately. 
I'm curious if a future OTA update would work.

**Comment by softhack007** [comment]
> I'm curious if a future OTA update would work.

Yes it should work. OTA can only replace the firmware but the bootloader and partition file that you have now will not be replaced. The OTA partition is about 2MB now, this should be good for longer time.

**Comment by netmindz** [comment]
Sorry for the slow reply. I can't find where I've put my matrix board at the moment. 

I have only flashed using platformio and then using the OTA I think.

It's a long time ago since I first got the board so it's possible I'm forgetting a step.

Using a regular S3 build from web installer then OTA I would expect to work.

I possibly needed to press one or more buttons actually, but I don't recall if that was for the first flash or recovery from a soft brick while doing the initial development work

**Comment by fonix232** [comment]
I've had this exact same issue on the MPS3, and practically none of the listed solutions worked:

- tried using the official WLED flasher (https://install.wled.me/) - wouldn't boot
- tried using the unofficial WLED web flasher (https://wled-install.github.io/) - wouldn't boot
- tried using the ESPHome web installer (https://web.esphome.io/) - wouldn't boot
- tried using the desktop WLED flasher (https://mm.kno.wled.ge/basics/install-wled-flasher/) - wouldn't boot
- tried ESPTool.js (https://espressif.github.io/esptool-js/) - wouldn't boot

On each method, I tried both directly flashing the WLEDMM MatrixPortal S3 build, as well as a generic WLED ESP32-S3 build, to no avail.

What DID work was using the basis of the ESPHome flasher - [ESP Web Tools](https://esphome.github.io/esp-web-tools/). This bad boy successfully flashed the stock S3 build, from which I managed to OTA update to the MatrixPortal-specific build.

I've also managed to repeat this process without any issues, while the other tools, once the flash was wiped (and many of them completely wipes the flash at the beginning of the process), would result in a non-booting unit.

**Comment by letmoc** [comment]
@softhack007 
Hello, I have the following problem with the board: it suddenly reboots. Attached is a screenshot. I am just running Akemi for testing. Then, after a few hours, sometimes even over 24 hours, it reboots.

![Screenshot-WLED-MM](https://github.com/user-attachments/assets/e544f7b7-1629-4da7-80f9-908741f74cc3)

**Comment by softhack007** [comment]
> @softhack007
> Hello, I have the following problem with the board: it suddenly reboots.

@letmoc I don't have the adafruit board, but i could dig out a similar board(S3 with qspi psram) and let it run for a day to check if i can reproduce the crash - "sw error" is not very specific, it could be anything - out-of-memory, bug in the code, bug in the HUB75 driver, network error, etc.

* How did you configure the led panel (screenshot from LEDs settings, please) ?
* does it also happen when running the board alone, without the panel connected ? 
   you can simply disconnect the HUB75 cable, as the software does not care if anything is connected to the HUB75 port or not.

**Comment by letmoc** [comment]
@softhack007 
I have two boards, and I updated one matrix portal to version 14.5.1-dev build 2501030. It has been running for exactly 24 hours. For testing, I have two different 64x64 displays with different driver ICs. Both behave the same. Attached are the screenshots.

![Screenshot-WLED-MM-2dsetup](https://github.com/user-attachments/assets/77e36691-c598-4c02-a03a-19f5786a6387)
![Screenshot-WLED-MM-hardware](https://github.com/user-attachments/assets/54c83e5c-ed94-4390-a694-6d6c018103da)

**Comment by softhack007** [comment]
> I have two different 64x64 displays with different driver ICs. Both behave the same

Sorry if I repeat myself - please disconnect the panel and let one board run "headless". It will help me to understand if the issue is reproducible when minimizing electrical "disturbances" created by the panels.

**Comment by letmoc** [comment]
no problem, remove the panel. give feedback

**Comment by letmoc** [comment]
@fonix232 
> What DID work was using the basis of the ESPHome flasher - [ESP Web Tools](https://esphome.github.io/esp-web-tools/). This bad boy successfully flashed the stock S3 build, from which I managed to OTA update to the MatrixPortal-specific build.

I can confirm it works perfectly like this, dear WLED friend. :-)

**Comment by softhack007** [comment]
> I can confirm it works perfectly like this, dear WLED friend. :-)

If you're happy with it, that's OK; ESP Web tools is also used by the official WLED web installer. 

### Just keep in mind you have been warned about *ESPhome Flasher*.

It might be that a few months after using this tool, you'll  see are crazy errors (e.g. cannot save preset or OTA update a new release errors out).

I'll simply dismiss any problem report after you used ESPHome Flasher. Same about the other wild ideas of first installing Tasmota, then OTA update with a WLED binary.

**Comment by softhack007** [comment]
> I can confirm it works perfectly like this, dear WLED friend. :-)

@fonix232  @letmoc  If you're happy with it, that's OK; ESP Web tools is also used by the official WLED web installer. 

### Just keep in mind you have been warned about *ESPhome Flasher*.

It might be that a few months after using this tool, you'll  see crazy errors (e.g. cannot save presets, preset or cfg corrupted,  or OTA update to a new release errors out).

I'll simply dismiss any problem report after you used ESPHome Flasher, because the tool causes nothing but trouble in user support.

Same about the other wild ideas of first installing Tasmota, then OTA update with a WLED binary.

**Comment by softhack007** [comment]
If you have a matrixportal board, I strongly recommend that you use the bootloader and tools from adafruit. Procedure is described [above](https://github.com/MoonModules/WLED/issues/205#issuecomment-2553617390). 

After initially flashing the adafruit firmware, then uploading a special partitions.bin and WLEDMM using [adafruit web flasher](https://adafruit.github.io/Adafruit_WebSerial_ESPTool), you can safely use OTA for updating to any WLED bin. 

It's just important that you pick a firmware for esp32-S3 with "qspi" psram.

**Comment by letmoc** [comment]
The test was conducted without a matrix display. After a little more than 48 hours, the restart occurred.

The second test with the board that I flashed using the method from fonix232 is currently running for 2 days and 7 hours without a restart.

**Comment by letmoc** [comment]
@softhack007 

The second module also rebooted. Just for your information.

**Comment by softhack007** [comment]
https://github.com/MoonModules/WLED/commit/46f41e90381f6512b9d810a656d5cca81cea8312

@letmoc I still did not find time to reproduce your problem. As it takes quite long until the crash occurs, my "educated guess" would be something like this
* memory leak
* something related with ntp (network time) updating local clocks on  the board
* something related to the wifi router re-newing the IP address of your board (DHCP). In this case, it might help to disable mDNS in WLED (wifi settings, clear the "mDNS address" field).

**Comment by netmindz** [comment]
Can we open a fresh issue please to track reboots and then we can close this installation one as resolved

**Comment by letmoc** [comment]
@netmindz @softhack007 
I have created a new topic; the discussion is now continued under https://github.com/MoonModules/WLED-MM/issues/222

**Comment by netmindz** [comment]
Are we now good to close this issue?

**Comment by Sireone** [comment]
After following the reset step from @softhack007 above & flashing both files, I still don't see any WLED-AP.  Thoughts? 

<img width="952" height="742" alt="Image" src="https://github.com/user-attachments/assets/a0c831e8-89bb-44d4-8500-33bc099bbf82" />

**Comment by softhack007** [comment]
@Sireone did you also perform this step

> follow adafruit instructions to install and run the factory-reset-and-bootloader firmware: https://learn.adafruit.com/adafruit-matrixportal-s3/factory-reset#factory-reset-and-bootloader-repair-3107941, and restart your board afterwards.
(use the adafruit web flasher aka "WebSerial ESPTool Method")

before running the adafruit web flasher? Its necessary to restore the adafruit special bootloader. If you don't have the adafruit bootloader, nothing will work.

Also keep your computer connected via USB and watch messages on serial monitor when starting WLED-MM on the matrixportal - without this we can't provide help.

**Comment by softhack007** [comment]
Update: adafruit official guide for installing WLED-MM on the matrixportal

https://learn.adafruit.com/matrix-portal-stained-glass-with-wled/install-wled-moonmodules

After first installation using the "adafruit way", use OTA update for upgrading to the latest release or nightly 
(there are firmware bin files for the matrixportal on our release page)

**Comment by abeumer** [comment]
> Ok, I did some research. It seems that adafruit wants it "their way", so its best to use their tools and use their stock bootloader.
> 
> 1. follow adafruit instructions to install and run the factory-reset-and-bootloader firmware: https://learn.adafruit.com/adafruit-matrixportal-s3/factory-reset#factory-reset-and-bootloader-repair-3107941
>    (use the adafruit web flasher aka "WebSerial ESPTool Method")
> 2. unzip the attached files, and flash them with the adafruit web flasher https://adafruit.github.io/Adafruit_WebSerial_ESPTool
>    
>    * offset `0x8000` partitions_8MB_tinyuf2.bin
>    * offset `0x10000` WLEDMM_14.5.0-beta_matrixportal_esp32s3_tinyuf2.bin
>    * hit "Program"
> 3. restart your matrixportal board. This should give you a working WLED-MM installation. At first boot, it may take 1-3 minutes before WLED-AP appears, because WLED needs some time to prepare the flash filesystem.
> 
> "should" because I don't have the matrixportal board to test.
> 
> 
> --> After this procedure (all 3 steps, no cutting corners), you _should_ be able to OTA update to any (newer) version of WLED or WLED-MM.
> 
> [matrixportal_tinyUF2.zip](https://github.com/user-attachments/files/18197561/matrixportal_tinyUF2.zip)

I'm running into the same issue trying to flash the S3 and the Adafruit WebSerial ESPTool gives me a error:
`Error: MD5 of file does not match data in flash!`

<img width="1461" height="754" alt="Image" src="https://github.com/user-attachments/assets/cc1626a1-8695-40c4-a95c-5ce9f3805738" />

**Comment by Emstar** [comment]
Just got my MatrixPortal and ended up here after I had flashing problems, thanks to all of you for sharing the solutions. The normal esp32 flashing routine sure wasn't working!

I wanted to add that once you have flashed the bin file that is part of this page with their tool, or with `esptool` at `0x0000` ...

https://learn.adafruit.com/matrix-portal-stained-glass-with-wled/install-wled-moonmodules

You can then flash new WLED MatrixPortal builds at `0x10000`. 

In other words, starting with that specific image they cooked up for that page can apparently let you avoid the factory reset procedure.
